onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+pair.  It should take `m` and a single `(x, ylabel)` pair as inputs and return
+the loss.
+
+Then write a function for calculating the LSE loss for a whole dataset.  It
+should take `m` and a list of `(x, ylabel)` pairs and return the LSE loss.
+
+Is it possible to write the second function by making multiple calls to the
+first function and combining the results?  Why or why not?
+
+## Minimizing the loss
+
+We have the loss, but what do we do to minimize it?  The answer is called
+"Gradient Descent".  We basically follow the curve of our loss until we get to
+the minimum of the loss graph:
+
+![gradient descent](gradient_descent.jpg "Gradient Descent")
+
+That image is pulled from from [this youtube video series](
+https://www.youtube.com/watch?v=b4Vyma9wPHo) which explains a lot of the
+details here in greater detail, if you want that.
+
+The basic idea is that we look at the slope of the loss curve (or the "gradient"
+of the loss curve if our loss is more than two dimensional), and try to take a
+step in the direction of the minimum.
+
+The direction of our step is based on the sign of the slope.  If the slope is
+positive, we know the minimum must be to our left, so we move in the negative
+direction (decreasing `m`).
+
+The size of our step is based on the magnitude of the slope.  If the slope is
+really steep, we proably can afford to take a really big step, but if it's
+really small we think we're getting close to the slope and we only take a small
+step.
+
+So let's define the step size `mstep` as:
+
+    mstep = - gradient * lr
+
+where `lr` is a constant multiplier, commonly called the "learning rate".
+
+But where does that `gradient` term come from?
+
+## The Gradient
+
+In the case of our simple model, we can use calculus to calculate the slope of
+our loss with respect to `m`, which for a one-variable model is also the
+gradient of the loss.
+
+We'll start by writing the loss for a single data point in terms of `m`:
+
+    loss = (ylabel - ypred)**2
+    loss = (ylabel - (m * x))**2
+
+Then apply the product rule and the chain rule from to get the derivative with
+respect to the weight `m`:
+
+    dloss/dm = - 2*x (ylabel - m*x)
+
+### Q3:
+
+Write a function to caclulate the gradient for a single data point.  It should
+take `m` and a single `(x, ylabel)` pair as inputs and return.
+
+Then write a function to calculate the total gradient for a list of
+`(x, ylabel)` pairs.
+
+Is it possible to write the second function by making multiple calls to the
+first function and combining the results?  Why or why not?
+
+## Autodifferentiation
+
+We solved our gradient symbolically because it was so simple.  But real
+gradients in real models are calculated with some magic called
+"autodifferentiation".
+
+You don't need to understand autodifferentiation, but you do need to
+understand:
+- it starts in the the "forward" pass of training a batch, and finishes in
+  the "backward" pass, and
+- it is additive.
```

Comment:
Basically, autodifferentiation:

- starts in the "forward" pass of training a batch and finishes in the "backward" pass, and
- is additive.

Reply:
I think the whole autodiff section is getting yanked, stuck in a new document, and rewritten